### DISCUSSION TITLE
Add option to exclude a File/Class/Function from layer dependency assertion

### DISCUSSION THREAD

**Discussion by f-rozanski** [LemonAppDev/konsist#566] [Request a Feature]
### Tell us more.

In tests, I would like to have less strict architecture tests than in production code. It will be great to have the option to exclude some files/classes or functions from layer dependency assertion. Let's assume I have a class called `XYZ.kt` in `infrastructure` package and a class called `ABC.kt` in `domain` package. `ABC.kt` uses `XYZ.kt` class thus there is an import to `XYZ.kt` and therefore `assertArchitecture` is failing. 
```
val testScope = Konsist.scopeFromTest()
testScope.assertArchitecture {
    val domain = Layer("Domain", "$APP_PACKAGE.domain..")
    val application = Layer("Application", "$APP_PACKAGE.application..")
    val infrastructure = Layer("Infrastructure", "$APP_PACKAGE.infrastructure..")

    domain.dependsOnNothing()
    application.dependsOn(domain)
    infrastructure.dependsOn(domain)
}
```
I tried to exclude `XYZ.kt`file from `testScope` but assertArchitecture` is still failing
```
val excludeScope = Konsist.scopeFromFile("../XYZ.kt")
val testScope = Konsist.scopeFromTest().minus(excludeScope)
testScope.assertArchitecture {
    val domain = Layer("Domain", "$APP_PACKAGE.domain..")
    val application = Layer("Application", "$APP_PACKAGE.application..")
    val infrastructure = Layer("Infrastructure", "$APP_PACKAGE.infrastructure..")

    domain.dependsOnNothing()
    application.dependsOn(domain)
    infrastructure.dependsOn(domain)
}
```
Do you have plans to add this feature? Or this feature already exist but I'm doing something wrong? In ArchUnit it is possible using `ImportOption`
```
import com.tngtech.archunit.core.importer.ClassFileImporter
import com.tngtech.archunit.core.importer.ImportOption
import com.tngtech.archunit.core.importer.Location
import com.tngtech.archunit.lang.syntax.ArchRuleDefinition.classes
import com.tngtech.archunit.library.Architectures.onionArchitecture

class ExcludeUtilityClasses : ImportOption {
    override fun includes(location: Location): Boolean =
        !listOf("XYZKt").any { location.contains(it) }
}
    
val importedClasses = ClassFileImporter()
    .withImportOption(ImportOption.Predefined.DO_NOT_INCLUDE_JARS)
    .withImportOption(ImportOption.Predefined.DO_NOT_INCLUDE_TESTS)
    .withImportOption(ExcludeUtilityClasses())
    .importPackages(APP_PACKAGE)
    
val hexagonalRule = onionArchitecture()
    .domainModels("..domain..")
    .domainServices("..domain..")
    .applicationServices("..application..")
    .adapter("infra", "..infrastructure..")

hexagonalRule.check(importedClasses)    
```

**Comment by igorwojda** [comment]
This is an interesting case. We will take a look.

In the future, we will consider adding architecture to assert the list of files, so filtering will be also possible this way:
```files
testScope
    .files // List<KoFileDeclaration>
    .withoutName("XYZ.kt")
    .assertArchitecture { // Not possible now

    }
```

**Reply by igorwojda** [reply]
BTW You can run `excludeScope.print()` to make sure this scope contains the file to be excluded.

**Reply by igorwojda** [reply]
Did it work for you or do you still think we have a bug when subtracting scope files? (If you think there is a bug perhaps you could help us a bit and share a simple project to reproduce the issue)

**Reply by igorwojda** [reply]
@f-rozanski can you elaborate on the above?

**Comment by f-rozanski** [comment]
Thanks for the answer 🙏 
> In the future, we will consider adding architecture to assert the list of files
Do you have any timeline for adding this feature? :)

**Reply by igorwojda** [reply]
No timeline yet. 
If you would like to contribute I can help you to kick this off.

**Reply by f-rozanski** [reply]
Maybe during hacktoberfest I could try. Could you estimate how much time it could take to make such a feature for a new person in this repo?

**Reply by igorwojda** [reply]
This should be quite straight forward. One day should be more then enough (if you don't hit any blockers).

**Comment by igorwojda** [comment]
A new version v0.13.0 has been released. It contains the fix.
See https://github.com/LemonAppDev/konsist/releases/tag/v0.13.0

// Now files can be further processed before asserting architecture
```kotlin
Konsist
  .scopeFromProject()
  .files
  .filter { ... } // remove file
  .assertArchitecture { ... } 
```

**Reply by f-rozanski** [reply]
It is amazing that it was released. To be honest I had plans to do this feature on the 18 Oct - my company selected that day for open source contributions due to Hacktoberfest. Now I don't have a task to pick 😆 but I'm sure I want to contribute to Konsist. What is the best way to communicate with you to pick a task and do some kickoff? On Twitter you have disabled DM's, on LinkedIn, I don't have a premium account to reach you. I requested access to Slack Kotlinlang but I'm waiting for approval.

**Reply by igorwojda** [reply]
What can I tell you, you are not the first person who told us that we are moving "too fast" 🤣

Bets place to chat with me and kickoff features is the official Kotlin Slack workspace
https://slack-chats.kotlinlang.org/ (DM)

We also have a dedicated Konsist channel there
https://kotlinlang.slack.com/archives/C05QG9FD6KS

This is one of the features that requires some polishing
https://lemonappdev.atlassian.net/browse/KON-106

Drop me there to Say Hi and we can have a kick off.

**Reply by igorwojda** [reply]
BTW @f-rozanski Please send me your email address on Slack. I will cerate an account for you in the JIRA, to assign the tickets and allow commenting
